plugins/obs-webrtc/whip-output.cpp

```diff
@@ -262,240 +342,465 @@ bool WHIPOutput::Setup()
+	// 2. It has credentials (implying it's a TURN server)
+	if (!url.empty() && (foundRel || hasCredentials)) {
+		try {
+			auto server = rtc::IceServer(url);
+			if (!username.empty() && !password.empty()) {
+				if (!hasCredentialType) {
+					do_log(LOG_WARNING, "ICE server has credentials but missing credential-type=password");
+				}
+				// We'll still try to use the credentials even if credential-type is missing
+				// as some implementations might omit it
+				server.username = username;
+				server.password = password;
+			}
+			iceServers.push_back(server);
+			do_log(LOG_INFO, "Added ICE server: %s", url.c_str());
+		} catch (const std::invalid_argument &err) {
+			do_log(LOG_WARNING, "Failed to parse ICE server: %s", err.what());
 		}
+	}
+}
 
-		if (pos == std::string::npos) {
-			break;
-		}
-		val.erase(0, pos + 1);
+void WHIPOutput::OnIceCandidate(const std::string &candidate, const std::string &mid) {
+	if (!running)
+		return;
+		
+	do_log(LOG_INFO, "Got ICE candidate: %s", candidate.c_str());
+	
+	if (trickle_endpoint.empty()) {
+		do_log(LOG_WARNING, "No trickle endpoint available");
+		return;
 	}
+	
+	SendTrickleCandidate(candidate);
+}
 
-	try {
-		auto iceServer = rtc::IceServer(url);
-		iceServer.username = username;
-		iceServer.password = password;
-		iceServers.push_back(iceServer);
-	} catch (const std::invalid_argument &err) {
-		do_log(LOG_WARNING, "Failed to construct ICE Server from %s: %s", val.c_str(), err.what());
+bool WHIPOutput::SendTrickleCandidate(const std::string &candidate) {
+	if (trickle_endpoint.empty()) {
+		do_log(LOG_DEBUG, "No trickle endpoint available");
+		return false;
 	}
+		
+	std::stringstream sdp;
+	sdp << "a=" << candidate << "\r\n";
+	
+	CURL *curl = curl_easy_init();
+	if (!curl) {
+		do_log(LOG_WARNING, "Failed to initialize CURL for trickle candidate");
+		return false;
+	}
+	
+	struct curl_slist *headers = nullptr;
+	if (!bearer_token.empty()) {
+		std::string auth = "Authorization: Bearer " + bearer_token;
+		headers = curl_slist_append(headers, auth.c_str());
+	}
+	
+	headers = curl_slist_append(headers, "Content-Type: application/trickle-ice-sdpfrag");
+	
+	curl_easy_setopt(curl, CURLOPT_URL, trickle_endpoint.c_str());
+	curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PATCH");
+	curl_easy_setopt(curl, CURLOPT_POSTFIELDS, sdp.str().c_str());
+	curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+	curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
+	
+	CURLcode res = curl_easy_perform(curl);
+	
+	curl_slist_free_all(headers);
+	curl_easy_cleanup(curl);
+	
+	return res == CURLE_OK;
 }
 
-bool WHIPOutput::Connect()
-{
+bool WHIPOutput::Connect() {
+	std::vector<rtc::IceServer> iceServers;
+	std::string read_buffer;
+	std::vector<std::string> http_headers;
+	
+	if (!peer_connection || !peer_connection->localDescription()) {
+		do_log(LOG_INFO, "No peer connection or local description available");
+		return false;
+	}
+
+	// Wait for ICE gathering to complete
+	int gather_timeout = 0;
```

Comment:
Why block for gathering? Would prefer not to do this, maybe I can help fix what is happening on this.